lib/container.ts

```diff
@@ -207,55 +207,114 @@ export class Container extends (EventEmitter as {
-	): AddOperation[] {
-		return this.getOperations(files, actionGroup);
+	): Promise<AddOperation[]> {
+		return await this.getOperations(files, actionGroup);
 	}
 
-	private getDeleteOperations(
+	private async getDeleteOperations(
 		files: string[],
 		actionGroup: DockerfileActionGroup,
-	): DeleteOperation[] {
+	): Promise<DeleteOperation[]> {
 		// for every file that we need to delete, find out where it was sent to
 		// and return the in-container path
-		return _.map(this.getOperations(files, actionGroup), 'toPath');
+		return _.map(await this.getOperations(files, actionGroup), 'toPath');
 	}
 
-	private getOperations(
+	private async getOperations(
 		files: string[],
 		actionGroup: DockerfileActionGroup,
-	): Array<{ fromPath: string; toPath: string }> {
-		return files.map(f => {
-			const matchingDep = Dockerfile.getActionGroupFileDependency(
-				f,
-				actionGroup,
-			);
-
-			/* istanbul ignore next */
-			if (matchingDep == null) {
-				throw new InternalInconsistencyError(
-					'Could not find matching file for action group',
+	): Promise<Array<{ fromPath: string; toPath: string }>> {
+		return Promise.all(
+			files.map(async f => {
+				const matchingDep = Dockerfile.getActionGroupFileDependency(
+					f,
+					actionGroup,
 				);
-			}
 
-			// TODO: I'm not sure the logic here is actually correct,
-			// specifically the way we build the destination when the containerPath
-			// is a directory
-			const strippedPath = matchingDep.sourceIsDirectory
-				? f
-				: f.split(Path.sep).pop()!;
-
-			const toPath = matchingDep.destinationIsDirectory
-				? Path.join(matchingDep.containerPath, strippedPath)
-				: matchingDep.containerPath;
-
-			return {
-				fromPath: f,
-				toPath,
-			};
+				/* istanbul ignore next */
+				if (matchingDep == null) {
+					throw new InternalInconsistencyError(
+						'Could not find matching file for action group',
+					);
+				}
+
+				// TODO: I'm not sure the logic here is actually correct,
+				// specifically the way we build the destination when the containerPath
+				// is a directory
+
+				// The dockerfile class tries to detect if the destination is a directory,
+				// but we can do one better with the actual container, and look directly
+				// at the destination
+				if (!matchingDep.destinationIsDirectory) {
+					matchingDep.destinationIsDirectory = await this.containerPathIsDirectory(
+						matchingDep.containerPath,
+					);
+				}
+
+				const strippedPath = matchingDep.sourceIsDirectory
+					? f
+					: f.split(Path.sep).pop()!;
+
+				const toPath = matchingDep.destinationIsDirectory
+					? Path.join(matchingDep.containerPath, strippedPath)
+					: matchingDep.containerPath;
+
+				return {
+					fromPath: f,
+					toPath,
+				};
+			}),
+		);
+	}
+
+	private async containerPathIsDirectory(path: string) {
+		const output = await this.executeCommandInternal([
+			'/bin/sh',
+			'-c',
+			`if [ -d "${path}" ]; then echo 'true'; else echo 'false'; fi`,
```

Comment:
Nit: if `executeCommandInternal` also returned the exit code, you could use `/usr/bin/test -d path`